cluster-autoscaler/core/scaledown/actuation/drain.go

```diff
@@ -19,128 +19,82 @@ package actuation
-	// is killed with SIGKILL after MaxGracefulTerminationTime
-	DefaultPodEvictionHeadroom = 30 * time.Second
-	// DefaultDsEvictionEmptyNodeTimeout is the time to evict all DaemonSet pods on empty node
-	DefaultDsEvictionEmptyNodeTimeout = 10 * time.Second
-	// DefaultDsEvictionRetryTime is a time between retries to create eviction that uses for DaemonSet eviction for empty nodes
-	DefaultDsEvictionRetryTime = 3 * time.Second
-)
-
 type evictionRegister interface {
 	RegisterEviction(*apiv1.Pod)
 }
 
-// Evictor can be used to evict pods from nodes.
-type Evictor struct {
-	EvictionRetryTime          time.Duration
-	DsEvictionRetryTime        time.Duration
-	DsEvictionEmptyNodeTimeout time.Duration
-	PodEvictionHeadroom        time.Duration
-	evictionRegister           evictionRegister
-	deleteOptions              simulator.NodeDeleteOptions
-}
-
-// NewDefaultEvictor returns an instance of Evictor using the default parameters.
-func NewDefaultEvictor(deleteOptions simulator.NodeDeleteOptions, evictionRegister evictionRegister) Evictor {
-	return Evictor{
-		EvictionRetryTime:          DefaultEvictionRetryTime,
-		DsEvictionRetryTime:        DefaultDsEvictionRetryTime,
-		DsEvictionEmptyNodeTimeout: DefaultDsEvictionEmptyNodeTimeout,
-		PodEvictionHeadroom:        DefaultPodEvictionHeadroom,
-		evictionRegister:           evictionRegister,
-		deleteOptions:              deleteOptions,
-	}
-}
-
-// DrainNode works like DrainNodeWithPods, but lists of pods to evict don't have to be provided. All non-mirror, non-DS pods on the
-// node are evicted. Mirror pods are not evicted. DaemonSet pods are evicted if DaemonSetEvictionForOccupiedNodes is enabled, or
-// if they have the EnableDsEvictionKey annotation.
-func (e Evictor) DrainNode(ctx *acontext.AutoscalingContext, nodeInfo *framework.NodeInfo) (map[string]status.PodEvictionResult, error) {
-	dsPodsToEvict, nonDsPodsToEvict := podsToEvict(ctx, nodeInfo)
-	return e.DrainNodeWithPods(ctx, nodeInfo.Node(), nonDsPodsToEvict, dsPodsToEvict)
-}
-
-// DrainNodeWithPods performs drain logic on the node. Marks the node as unschedulable and later removes all pods, giving
-// them up to MaxGracefulTerminationTime to finish. The list of pods to evict has to be provided.
-func (e Evictor) DrainNodeWithPods(ctx *acontext.AutoscalingContext, node *apiv1.Node, pods []*apiv1.Pod, daemonSetPods []*apiv1.Pod) (map[string]status.PodEvictionResult, error) {
+// drainNodeWithPodsBasedOnPodPriority performs drain logic on the node based on pod priorities. Marks the node as unschedulable and
+// later removes all pods, giving them up to MaxGracefulTerminationTime to finish. The list of pods to evict has to be provided.
+func drainNodeWithPodsBasedOnPodPriority(ctx *acontext.AutoscalingContext, node *apiv1.Node, pods []*apiv1.Pod, shutdownGracePeriodByPodPriority []kubelet_config.ShutdownGracePeriodByPodPriority,
```

Comment:
The extracted logic is very hard to read, I think it might be too attached to the previous flow/terminology that doesn't make sense in isolation from the rest of the eviction logic. Plus it was convoluted from the start, and adding the `ignoreDaemonSetPodEvictionResult` ifs everywhere just adds even more complexity.

WDYT about taking two separate lists of pods instead - e.g. `fullEvictionPods` and `bestEffortEvictionPods`, and treating the two separately instead of the current flow?

Reply:
Done